BUG(?) `Index.__getitem__` with 0D numpy array raises AssertionError

```python
In [5]: pd.Index([2, 2, 2])[np.array(0)]
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 pd.Index([2, 2, 2])[np.array(0)]

File ~/pandas/core/indexes/base.py:5370, in Index.__getitem__(self, key)
   5366     disallow_ndim_indexing(result)
   5368 # NB: Using _constructor._simple_new would break if MultiIndex
   5369 #  didn't override __getitem__
-> 5370 return self._constructor._simple_new(result, name=self._name)

File ~/pandas/core/indexes/base.py:641, in Index._simple_new(cls, values, name, refs)
    631 @classmethod
    632 def _simple_new(
    633     cls, values: ArrayLike, name: Hashable | None = None, refs=None
    634 ) -> Self:
    635     """
    636     We require that we have a dtype compat for the values. If we are passed
    637     a non-dtype compat, then coerce using the constructor.
    638 
    639     Must be careful not to recurse.
    640     """
--> 641     assert isinstance(values, cls._data_cls), type(values)
    643     result = object.__new__(cls)
    644     result._data = values

AssertionError: <class 'numpy.int64'>
```

Not sure if we should allow this case, but I would expect either `2` or a `ValueError`

--- a/doc/source/whatsnew/v3.0.0.rst
+++ b/doc/source/whatsnew/v3.0.0.rst
@@ -945,6 +945,7 @@ Indexing
 - Bug in :meth:`Series.__setitem__` when assigning boolean series with boolean indexer will raise ``LossySetitemError`` (:issue:`57338`)
 - Bug in printing :attr:`Index.names` and :attr:`MultiIndex.levels` would not escape single quotes (:issue:`60190`)
 - Bug in reindexing of :class:`DataFrame` with :class:`PeriodDtype` columns in case of consolidated block (:issue:`60980`, :issue:`60273`)
+- Bug in :meth:`Index.__getitem__` incorrectly raising with a 0-dim ``np.ndarray`` key (:issue:`55601`)
 
 Missing
 ^^^^^^^

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -5219,6 +5219,7 @@ class Index(IndexOpsMixin, PandasObject):
         """
         getitem = self._data.__getitem__
 
+        key = lib.item_from_zerodim(key)
         if is_integer(key) or is_float(key):
             # GH#44051 exclude bool, which would return a 2d ndarray
             key = com.cast_scalar_indexer(key)

--- a/pandas/core/indexes/multi.py
+++ b/pandas/core/indexes/multi.py
@@ -2235,6 +2235,7 @@ class MultiIndex(Index):
     # --------------------------------------------------------------------
 
     def __getitem__(self, key):
+        key = lib.item_from_zerodim(key)
         if is_scalar(key):
             key = com.cast_scalar_indexer(key)
 

--- a/pandas/core/indexes/range.py
+++ b/pandas/core/indexes/range.py
@@ -1175,6 +1175,7 @@ class RangeIndex(Index):
         """
         Conserve RangeIndex type for scalar and slice keys.
         """
+        key = lib.item_from_zerodim(key)
         if key is Ellipsis:
             key = slice(None)
         if isinstance(key, slice):

--- a/pandas/tests/indexes/test_any_index.py
+++ b/pandas/tests/indexes/test_any_index.py
@@ -117,6 +117,15 @@ class TestRoundTrips:
 
 
 class TestIndexing:
+    def test_getitem_0d_ndarray(self, index):
+        # GH#55601
+        if len(index) == 0:
+            pytest.skip(reason="Test assumes non-empty index")
+        key = np.array(0)
+        result = index[key]
+
+        assert result == index[0]
+
     def test_get_loc_listlike_raises_invalid_index_error(self, index):
         # and never TypeError
         key = np.array([0, 1], dtype=np.intp)

